File: scripts/bindings/hifi.dart
class AACube {
	num x;
	num y;
	num z;
	num scale;
}

class Vec3 {
	num x;
	num y;
	num z;
}

class AttachmentData {
	String modelUrl;
	String jointName;
	Vec3 translation;
	Vec3 rotation;
	num scale;
	bool soft;
}

class Collision {
	num type;
	Vec3 penetration;
	Vec3 contactPoint;
	Vec3 velocityChange;
}

class Color {
	int red;
	int green;
	int blue;
}

class Quat {
	num x;
	num y;
	num z;
	num w;
}

class FBXAnimationFrame {
	List<Quat> rotations;
	List<Vec3> translations;
}

class KeyEvent {
	int key;
	String text;
	bool isShifted;
	bool isMeta;
	bool isControl;
	bool isAlt;
	bool isKeypad;
	bool isAutoRepeat;
}

class Vec3Color {
	int x;
	int y;
	int z;
}

class Material {
	String name = "";
	String model = "hifi_pbr";
	Vec3Color emissive;
	double opacity = 1.0;
	bool unlit = false;
	Vec3Color albedo;
	double roughness;
	double metallic;
	double scattering;
	String emissiveMap;
	String albedoMap;
	String opacityMap;
	String roughnessM<ap;
	String glossMap;
	String metallicMap;
	String specularMap;
	String normalMap;
	String bumpMap;
	String occlusionMap;
	String scatteringMap;
	String lightMap;
}

class MaterialResource {
	int materialVersion = 1;
	dynamic materials;
}

class MouseEvent {
	int x;
	int y;
	String button;
	bool isLeftButton;
	bool isMiddleButton;
	bool isRightButton;
	bool isShifted;
	bool isMeta;
	bool isControl;
	bool isAlt;
}

class PickRay {
	Vec3 origin;
	Vec3 direction;
}

enum PickType {
	Ray, Stylus
}

class Vec2 {
	num x;
	num y;
}

class PointerEvent {
	String type;
	int id;
	Vec2 pos2D;
	Vec3 pos3D;
	Vec3 normal;
	Vec3 direction;
	String button;
	bool isPrimaryButton;
	bool isLeftButton;
	bool isSecondaryButton;
	bool isRightButton;
	bool isTertiaryButton;
	bool isMiddleButton;
	bool isPrimaryHeld;
	bool isSecondaryHeld;
	bool isTertiaryHeld;
	int keyboardModifiers;
}

class Pose {
	Vec3 translation;
	Quat rotation;
	Vec3 velocity;
	Vec3 angularVelocity;
	bool valid;
}

class RayPickResult {
	int type;
	bool intersects;
	double distance;
	Vec3 intersection;
	Vec3 surfaceNormal;
	dynamic extraInfo;
	PickRay searchRay;
}

class Rect {
	int x;
	int y;
	int width;
	int height;
}

class Size {
	num height;
	num width;
}

class StylusTip {
	int side;
	Vec3 position;
	Quat orientation;
	Vec3 velocity;
}

class StylusPickResult {
	int type;
	bool intersects;
	double distance;
	Vec3 intersection;
	Vec3 surfaceNormal;
	dynamic extraInfo;
	StylusTip stylusTip;
}

class TouchEvent {
	int x;
	int y;
	bool isPressed;
	bool isMoved;
	bool isStationary;
	bool isReleased;
	bool isShifted;
	bool isMeta;
	bool isControl
	bool isAlt;
	int touchPoints;
	List<Vec2> points;
	double radius;
	bool isPinching;
	bool isPinchOpening;
	double angle;
	double deltaAngle;
	List<double> angles;
	bool isRotating;
	String rotating;
}

class Vec4 {
	num x;
	num y;
	num z;
	num w;
}

class ViewFrustrum {
	Vec3 position;
	Quat orientation;
	double centerRadius;
	double fieldOfView;
	num aspectRatio;
}

class WheelEvent {
	int x;
	int y;
	int delta;
	String orientation;
	bool isLeftButton;
	bool isMiddleButton;
	bool isRightButton;
	bool isShifted;
	bool isMeta;
	bool isControl;
	bool isAlt;
}

class AudioEffectOptions {
	num bandwidth = 10000;
	num preDelay = 20;
	num lateDelay = 0;
	num reverbTime = 2;
	num earlyDiffusion = 100;
	num lateDiffusion = 100;
	num roomSize = 50;
	num density = 100;
	double bassMult = 1.5;
	num bassFreq = 250;
	num highGain = -6;
	num highFreq = 3000;
	double modRate = 2.3;
	num modDepth = 50;
	num earlyGain = 0;
	num lateGain = 0;
	num earlyMixLeft = 20;
	num earlyMixRight = 20;
	num lateMixLeft = 90;
	num lateMixRight = 90;
	num wetDryMix = 50;
}

class KeyLight {
	Vec3 color;
	num intensity;
	num ambientIntensity;
	Vec3 direction;
}

class Location {
	num longitude;
	num latitude;
	num altitude;
}

class Time {
	int hour;
	int day;
}

File: scripts/bindings/nim/hifi.nim
type
  AACube* = ref object
    x*: float
    y*: float
    z*: float
    scale*: float

  Vec3* = ref object
    x*: float
    y*: float
    z*: float

  AttachmentData* = ref object
    modelUrl*: cstring
    jointName*: cstring
    translation*: Vec3
    rotation*: Vec3
    scale*: float
    soft*: bool

  Uuid* = ref object

  BoxFace*{.pure.} = enum
    MIN_X_FACE = "MIN_X_FACE"
    MAX_X_FACE = "MAX_X_FACE"
    MIN_Y_FACE = "MIN_Y_FACE"
    MAX_Y_FACE = "MAX_Y_FACE"
    MIN_Z_FACE = "MIN_Z_FACE"
    MAX_Z_FACE = "MAX_Z_FACE"
    UNKNOWN_FACE = "UNKNOWN_FACE"

  Collision* = ref object
    idA*: Uuid
    idB*: Uuid
    penetration*: Vec3
    contactPoint*: Vec3
    velocityChange*: Vec3

  Color* = ref object
    red*: uint8
    green*: uint8
    blue*: uint8

  Quat* = ref object
    x*: float
    y*: float
    z*: float
    w*: float

  Vec4* = distinct Quat

  ContactEventType* = enum
    Start = 0
    Continuation = 1
    End = 2

  FBXAnimationFrame* = ref object
    rotations*: seq[Quat]
    translations*: seq[Vec3]

  KeyboardModifier*{.pure.} = enum
    Shift = 25
    Control = 26
    Alt = 27
    Meta = 28
    Keypad = 29
    Group = 30

  KeyboardModifiers* = set of KeyboardModifier

  KeyEvent* = ref object
    key*: int
    text*: cstring
    isShifted*: bool
    isMeta*: bool
    isControl*: bool
    isAlt*: bool
    isKeypad*: bool
    isAutoRepeat*: bool

  Vec3Color* = ref object
    x*: int
    y*: int
    z*: int

  Material* = ref object
    name*: cstring
    model*: cstring
    emissive*: Vec3Color
    opacity*: float
    unlit*: bool
    albedo*: Vec3Color
    roughness*: float
    metallic*: float
    scattering*: float
    emissiveMap*: cstring
    albedoMap*: cstring
    opacityMap*: cstring
    roughnessMap*: cstring
    glossMap*: cstring
    metallicMap*: cstring
    specularMap*: cstring
    normalMap* cstring
    bumpMap*: cstring
    occlusionMap*: cstring
    scatteringMap*: cstring
    lightMap*: cstring

  MaterialResource* = ref object
    materialVersion*: int
    materials*: Material | seq[Material]

  MouseEvent* = ref object
    x*: int
    y*: int
    button*: cstring
    isLeftButton*: bool
    isMiddleButton*: bool
    isRightButton*: bool
    isShifted*: bool
    isMeta*: bool
    isControl*: bool
    isAlt*: bool

  PickRay* = ref object
    origin*: Vec3
    direction*: Vec3

  PickType* = enum
    Ray, Stylus

  Vec2* = ref object
    x*: float
    y*: float

  PointerEvent* = ref object
    kind*: cstring
    id*: int
    pos2D*: Vec2
    pos3D*: Vec3
    normal*: Vec3
    direction*: Vec3
    button*: cstring
    isPrimaryButton*: bool
    isLeftButton*: bool
    isSecondaryButton*: bool
    isRightButton*: bool
    isTertiaryButton*: bool
    isMiddleButton*: bool
    isPrimaryHeld*: bool
    isSecondaryHeld*: bool
    isTertiaryHeld*: bool
    keyboardModifiers*: KeyboardModifiers

  Pose* = ref object
    translation*: Vec3
    rotation*: Quat
    velocity*: Vec3
    angularVelocity*: Vec3
    valid*: bool

  RayPickResult* = ref object
    kind*: int
    intersects*: bool
    objectID*: Uuid
    distance*: float
    intersection*: Vec3
    surfaceNormal*: Vec3
    extraInfo*: JsObject
    searchRay*: PickRay

  Rect* = ref object
    x*: int
    y*: int
    width*: int
    height*: int

  Size* = ref object
    height*: float
    width*: float

  StylusTip* = ref object
    side*: int
    position*: Vec3
    orientation*: Quat
    velocity*: Vec3

  StylusPickResult* = ref object
    kind*: int
    intersects*: bool
    objectID*: Uuid
    distance*: float
    intersection*: Vec3
    surfaceNormal*: Vec3
    extraInfo*: JsObject
    stylusTip*: StylusTip

  TouchEvent* = ref object
    x*: float
    y*: float
    isPressed*: bool
    isMoved*: bool
    isStationary*: bool
    isReleased*: bool
    isShifted*: bool
    isMeta*: bool
    isControl*: bool
    isAlt*: bool
    touchPoints*: int
    points*: seq[Vec2]
    radius*: float
    isPinching*: bool
    isPinchOpening*: bool
    angle*: float
    deltaAngle*: float
    angles*: seq[float]
    isRotating*: bool
    rotating*: cstring

  Mat4* = ref object

  ViewFrustrum* = ref object
    position*: Vec3
    orientation*: Quat
    centerRadius*: float
    fieldOfView*: float
    aspectRatio*: float
    projection*: Mat4

  WheelEvent* = ref object
    x*: float
    y*: float
    delta*: float
    orientation*: cstring
    isLeftButton*: bool
    isMiddleButton*: bool
    isRightButton*: bool
    isShifted*: bool
    isMeta*: bool
    isControl*: bool
    isAlt*: bool

  AnimationObject* = ref object
    jointNames*: seq[cstring]
    frames*: seq[FBXAnimationFrame]

  AudioEffectOptions* = ref object
    bandwidth*: float
    preDelay*: float
    lateDelay*: float
    reverbTime*: float
    earlyDiffusion*: float
    lateDiffusion*: float
    roomSize*: float
    density*: float
    bassMult*: float
    bassFreq*: float
    highGain*: float
    highFreq*: float
    modRate*: float
    modDepth*: float
    earlyGain*: float
    lateGain*: float
    earlyMixLeft*: float
    earlyMixRight*: float
    lateMixLeft*: float
    lateMixRight*: float
    wetDryMix*: float

  ReverbOptions* = AudioEffectOptions

  AudioStreamStats* = ref object
    lossRate*: float
    lossCount*: int
    lossRateWindow*: float

  MappingObject* = ref object

  MeshProxy* = ref object

  OverlayWebWindow* = ref object
    url: cstring
    position*: Vec2
    size*: Vec2
    visible*: bool

  Properties* = ref object
    title*: cstring
    source*: cstring
    width*: float
    height*: float
    visible*: bool

  OverlayWindow* = ref object
    position*: Vec2
    size*: Vec2
    visible*: bool

  State* = enum
    QUEUED
    LOADING
    LOADED
    FINISHED
    FAILED

  ResourceObject* = ref object
    url*: cstring
    state*: State

  RouteObject* = ref object

  KeyLight* = ref object
    color*: Vec3
    intensity*: float
    ambientIntensity*: float
    direction*: Vec3

  Location* = ref object
    longitude*: float
    latitude*: float
    altitude*: float

  Stage* = ref object
    backgroundMode* = cstring
    keyLight*: KeyLight
    location*: Location
    sunModel*: bool
    time*: Time

  ScriptsModel* = ref object

  ScriptsModelFilter* = ref object

  SoundObject* = ref object
    downloaded*: bool
    duration*: float

  ButtonProperties* = ref object
    icon*: cstring
    hoverIcon*: cstring
    activeHoverIcon*: cstring
    activeIcon*: cstring
    text*: cstring
    hoverText*: cstring
    activeText*: cstring
    activeHoverText*: cstring
    isActive*: bool
    sortOrder*: int

  TabletButtonProxy* = ref object
    uuid: Uuid
    properties*: ButtonProperties

  ButtonList* = ref object

  TabletProxy* = ref object
    name: cstring
    toolbarMode*: bool
    landscape*: bool
    tabletShown: bool
    buttons: ButtonList

  ToolbarButtonProxy* = ref object

  ToolbarProxy* = ref object

proc `buttons`*(self: TabletProxy): ButtonList =
  return self.buttons

proc `name`*(self: TabletProxy): cstring =
  return self.name

proc `tabletShown`*(self: TabletProxy): bool =
  return self.tabletShown

proc `url`*(self: OverlayWebWindow): cstring =
  return self.url

proc `uuid`*(self: TabletButtonProxy): Uuid =
  return self.uuid

proc disable*(self: MappingObject): MappingObject
  {.importcpp: "disable".}

proc enable*(self: MappingObject, doEnable: bool = true): MappingObject
  {.importcpp: "enable".}

proc getFrames*(self: AnimationObject): seq[FBXAnimationFrame]
  {.importcpp: "getFrames".}

proc getJointNames*(self: AnimationObject): seq[cstring]
  {.importcpp: "getJointNames".}

proc getNumVertices*(self: MeshProxy): int
  {.importcpp: "getNumVertices".}

proc getPos*(self: MeshProxy, index: int): Vec3
  {.importcpp: "getPos".}

proc getPosition*(self: OverlayWebWindow): Vec2
  {.importcpp: "getPosition".}

proc getSize*(self: OverlayWebWindow): Vec2
  {.importcpp: "getSize".}

proc getURL*(self: OverlayWebWindow): cstring
  {.importcpp: "getURL".}

proc initQml*(self: var OverlayWebWindow, properties: Properties)
  {.importcpp: "initQml".}

proc isVisible*(self: OverlayWebWindow): bool
  {.importcpp: "isVisible".}

proc setPosition*(self: var OverlayWebWindow, position: Vec2)
  {.importcpp: "setPosition".}

proc setPosition*(self: var OverlayWebWindow, x: int, y, int)
  {.importcpp: "setPosition".}

proc setScriptURL*(self: var OverlayWebWindow, script: cstring)
  {.importcpp: "setScriptURL".}

proc setSize*(self: var OverlayWebWindow, size: Vec2)
  {.importcpp: "setSize".}
